routing/payment_lifecycle_test.go

```diff
@@ -617,6 +617,40 @@ func TestRouterPaymentStateMachine(t *testing.T) {
 			},
 			paymentErr: channeldb.FailureReasonPaymentDetails,
 		},
+		{
+			// A MP payment scenario when our path finding returns
+			// after we've just received a terminal failure, and
+			// demonstrates a bug where the payment will return with
+			// and "unexpected" ErrPaymentAlreadyFailed rather than
+			// failing with a permanent error. This results in the
+			// payment getting stuck.
+			name: "MP path found after failure",
+
+			steps: []string{
+				routerInitPayment,
+
+				// shard 0
+				routeRelease,
+				routerRegisterAttempt,
+				sendToSwitchSuccess,
+
+				// The first shard fail with a terminal error.
+				getPaymentResultTerminalFailure,
+				routerFailAttempt,
+				routerFailPayment,
+
+				// shard 1
+				routeRelease,
+				routerRegisterAttempt,
```

Comment:
should add a comment here that we expect it to register the attempt, but this actually fails. Do we have a way of assersting here that the registration fails with the `ErrPaymentAlreadyFailed` error?

Reply:
This sounds like a good idea to me. I think this solution will cover us in the test, but I have a question as to whether this will cover this bug case "in the wild".

The definition of a "FAILED" payment looks a bit different in testing as compared to the live `resumePayment()` payment lifecycle machine. See the following:
- [Register Attempt (Mock)](https://github.com/lightningnetwork/lnd/blob/master/routing/mock_test.go#L294-L296)
- [Register Attempt (Live)](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payment_control.go#L288-L294)

Both check to see whether the top level payment is "FAILED" before allowing a new shard registration. The mock determines failure with a simple lookup in the `mockControlTower`'s `failed` map. However in the live case, when we [register a new shard with the ControlTower/DB](https://github.com/lightningnetwork/lnd/blob/master/routing/payment_lifecycle.go#L380), we fetch the payment and [ensure the payment is `IN-FLIGHT`](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payment_control.go#L288-L294). With the way [`fetchPayment()`](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payments.go#L358-L365) works, a payment is reported as `status=IN-FLIGHT` by default if it has any in-flight shards. Given this, I don’t see how we would be returning the [`ErrPaymentAlreadyFailed`](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payment_control.go#L670). If [`shard 1`](https://github.com/lightningnetwork/lnd/pull/5214#issue-615993831) is stuck `IN-FLIGHT`, it looks like `fetchPayment()` will tell us the payment is still in flight.

I think we might be hitting the [terminal info check](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payment_control.go#L300-L302) just below (which exists in the wild but not in our mock). This is a looser check which is sensitive to any payment level failure and does not care about whether there are in flight shards or not. In this case, it would not be `ErrPaymentAlreadyFailed`, but actually `ErrPaymentTerminal` which we're hitting in the wild.

Here’s a diff with what I think will get the mock of `RegisterAttempt` to be more inline with how it looks like it works in the wild: https://github.com/lightningnetwork/lnd/commit/4e388c00571d4aa69b61ff42f978016941dcfa91